aziot/src/check.rs

```diff
@@ -0,0 +1,326 @@
+
+    /// Output format. Note that JSON output contains some additional information
+    /// like OS name, OS version, disk space, etc.
+    #[structopt(short, long, value_name = "FORMAT", default_value = "text")]
+    output: OutputFormat,
+
+    /// Increases verbosity of output.
+    #[structopt(short, long)]
+    verbose: bool,
+
+    /// Treats warnings as errors. Thus 'aziot check' will exit with non-zero
+    /// code if it encounters warnings.
+    #[structopt(long)]
+    warnings_as_errors: bool,
+
+    #[structopt(flatten)]
+    checker_cfg: CheckerCfg,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum OutputFormat {
+    Text,
+    Json,
+    JsonStream,
+}
+
+impl FromStr for OutputFormat {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, &'static str> {
+        Ok(match s {
+            "text" => OutputFormat::Text,
+            "json" => OutputFormat::Json,
+            "json-stream" => OutputFormat::JsonStream,
+            _ => return Err("invalid output format"),
+        })
+    }
+}
+
+pub async fn check(mut cfg: CheckCfg) -> Result<()> {
+    // manually pass verbosity down to the checker-specific configuration
+    cfg.checker_cfg.verbose = cfg.verbose;
+    let cfg = cfg; // freeze cfg
+
+    let checker_shared = CheckerShared::new(cfg.checker_cfg);
+
+    let mut checks: BTreeMap<String, CheckOutputSerializable> = Default::default();
+    let mut check_data = crate::internal::check::all_checks();
+    let mut check_cache = CheckerCache::new();
+
+    let mut num_successful = 0_usize;
+    let mut num_warnings = 0_usize;
+    let mut num_skipped = 0_usize;
+    let mut num_fatal = 0_usize;
+    let mut num_errors = 0_usize;
+
+    macro_rules! output {
+        ($color:ident, $($args:tt)*) => {
+            if matches!(cfg.output, OutputFormat::Text) {
+                print!("{}", format!($($args)*).$color());
+            }
+        };
+    }
+
+    macro_rules! outputln {
+        () => {
+            if matches!(cfg.output, OutputFormat::Text) {
+                println!();
+            }
+        };
+        ($color:ident, $($args:tt)*) => {
+            if matches!(cfg.output, OutputFormat::Text) {
+                println!("{}", format!($($args)*).$color());
+            }
+        };
+    }
+
+    macro_rules! outputlns {
+        ($color:ident, $first_line_indent:expr, $other_line_indent:expr, $lines:expr $(,)?) => {
+            for (i, line) in $lines.enumerate() {
+                outputln!(
+                    $color,
+                    "{}{}",
+                    if i == 0 {
+                        $first_line_indent
+                    } else {
+                        $other_line_indent
+                    },
+                    line
+                );
+            }
+        };
+    }
+
+    'all_checks: for (section_name, section_checks) in &mut check_data {
+        outputln!(normal, "{}", section_name);
+        outputln!(normal, "{}", "-".repeat(section_name.len()));
```

Comment:
`outputln!(normal, "{:->section_name_len$}", "", section_name_len = section_name.len());`

... will do it without allocating a `String` like `str::repeat` does.

Reply:
Ah, yep, very true. I carbon-copied this formatting from `iotedge check`, so I wasn't really paying much attention.

I mean, it doesn't _really_ matter in this particular case, since we're not min-maxing for memory use, but sure, I'll change it.